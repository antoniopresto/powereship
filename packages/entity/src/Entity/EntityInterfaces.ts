import {
  ExtendDefinitionResult,
  GraphType,
  ObjectDefinitionInput,
  ObjectFieldInput,
  parseObjectDefinition,
  ToFinalField,
} from '@brabo/schema';
import { MaybeArray, UnionToIntersection } from '@brabo/utils';

import {
  AnyCollectionIndexConfig,
  CollectionConfigIndexes,
  CreateOneResult,
  DocumentBase,
  DocumentIndexItem,
  DocumentMethods,
  LoaderContext,
  ParsedDocumentIndexes,
  ParsedIndexKey,
  TransporterLoaderName,
} from '../Transporter';

import {
  EntityGraphQLConditionsType,
  EntityGraphQLFieldConditionsType,
} from './EntityFilterConditionType';
import {
  EntityDocFromType,
  EntityFieldResolver,
  EntityOptions,
} from './EntityOptions';
import { EntityHookOptions } from './EntityPlugin';
import { EntityOperationInfosRecord } from './entityOperationContextTypes';
import { EdgeType, PaginationType } from './paginationUtils';

export type EntityGeneratedFields = ReturnType<
  typeof createEntityDefaultFields
>;

export const createEntityDefaultFields = () =>
  _EntityGeneratedFields({
    createdAt: { type: 'date' },
    createdBy: {
      optional: true,
      type: 'string',
    },
    id: { type: 'string' },
    ulid: { type: 'ulid' },
    updatedAt: { type: 'date' },
    updatedBy: {
      optional: true,
      type: 'string',
    },
  });

type GetLoaderFilterDef<Loader, DocDef> = Loader extends (
  config: infer Config
) => any
  ? Config extends { filter: infer Filter }
    ? UnionToIntersection<Filter> extends infer AllFilter
      ? {
          [K in keyof AllFilter as K extends keyof DocDef
            ? K
            : never]: K extends keyof DocDef ? DocDef[K] : never;
        } extends infer R
        ? {
            // transforming optional args as optional field definitions
            [K in keyof R]-?: [Extract<R[K], undefined>] extends [never]
              ? ToFinalField<R[K]>
              : Omit<
                  ToFinalField<Exclude<R[K], undefined>>,
                  '__infer' | 'optional'
                > & { optional: true } extends infer F
              ? { [K in keyof F]: F[K] } & {}
              : never;
          }
        : never
      : never
    : never
  : never;

export type DefaultEntityFields = {
  createdAt: Date;
  createdBy: string | undefined;
  id: string;
  ulid: string;
  updatedAt: Date;
  updatedBy: string | undefined;
};

export type EntityFinalDefinition<InputDef> = InputDef extends {
  definition: { def: infer Definition };
}
  ? {
      [K in keyof EntityGeneratedFields as K extends keyof Definition
        ? never
        : K]: EntityGeneratedFields[K];
    } & {
      [K in keyof Definition]: Definition[K];
    } extends infer R
    ? {
        [K in keyof R]: R[K];
      }
    : never
  : never;

type _GetLoaderUtils<Loader, Type> = Loader extends unknown
  ? EntityFinalDefinition<Type> extends infer Def
    ? GetLoaderFilterDef<Loader, Def> extends infer Filter
      ? Filter extends unknown
        ? {
            filterDef: Filter;
            queryArgs: {
              after: {
                optional: true;
                type: 'ID';
              };
              condition: Type extends {
                definition: infer Def;
              }
                ? Def extends { def: infer Def }
                  ? {
                      def: EntityGraphQLFieldConditionsType<Def>;
                      optional: true;
                      type: 'object';
                    }
                  : never
                : never;
              filter: {
                def: Filter;
                type: 'object';
              };
              first: {
                optional: true;
                type: 'int';
              };
            };
          }
        : never
      : never
    : never
  : never;

interface PrimaryEntityMethods<Options extends EntityOptions> {
  conditionsDefinition: Options['type'] extends {
    definition: infer Def;
  }
    ? Def extends { def: infer Def }
      ? {
          def: EntityGraphQLConditionsType<Def>;
          type: 'object';
        }
      : never
    : never;
  edgeType: EdgeType<Options['type']>;
  getDocumentId(doc: Record<string, any>): string;

  indexGraphTypes: {
    [K in Options['indexes'][number]['name']]: GraphType<{
      object: ObjectDefinitionInput;
    }>;
  };

  indexes: Options['indexes'];

  inputDefinition: Options['type'] extends { definition: infer Def }
    ? Def extends { def: infer Def }
      ? {
          [K in keyof Def]: ToFinalField<Def[K]>;
        }
      : never
    : never;

  name: Options['name'];

  originType: Options['type'];
  paginationType: PaginationType<Options['type']>;

  parse: (
    ...args: Parameters<Options['type']['parse']>
  ) => EntityDocFromType<Options['type']>;

  parseDocumentIndexes(doc: Record<string, any>): ParsedDocumentIndexes;

  transporter: Options['transporter'];

  type: ((x: Options['type']) => any) extends (x: infer Type) => any
    ? GraphType<{ object: EntityFinalDefinition<Type> }>
    : never;

  updateDefinition: ((x: Options['type']) => any) extends (x: infer Type) => any
    ? Type extends { definition: { def: infer D } }
      ? {
          [K in keyof D]: ToFinalField<D[K]> extends infer R
            ? {
                [K in keyof R as K extends '__infer'
                  ? never
                  : K]: K extends 'optional' ? true : R[K];
              } & {}
            : never;
        }
      : never
    : never;
}

// export type EntityGraphType<T> = T extends
export type Entity<Options extends EntityOptions> =
  Options['type'] extends infer Type
    ? Type extends {
        definition: { def: unknown };
        parse(...args: unknown[]): unknown;
      }
      ? EntityDocFromType<Options['type']> extends infer Doc
        ? Doc extends DocumentBase
          ? Options['indexes'] extends infer Indexes
            ? Indexes extends CollectionConfigIndexes<Doc>
              ? PrimaryEntityMethods<Options> extends infer PropsFromConfig
                ? /**
                   * HADOUKEN!!!
                   */
                  EntityLoadersWithUtils<
                    Type,
                    Doc,
                    Indexes
                  > extends infer Loaders
                  ? {
                      [K in keyof Loaders]: _GetLoaderUtils<
                        Loaders[K],
                        Options['type']
                      > extends infer Utils
                        ? [Utils] extends [never]
                          ? Loaders[K]
                          : Loaders[K] & Utils
                        : never;
                    } extends infer LoaderWithUtils
                    ? {
                        [K in
                          | keyof PropsFromConfig
                          | keyof LoaderWithUtils
                          | 'loaders']: K extends keyof LoaderWithUtils
                          ? LoaderWithUtils[K]
                          : K extends 'loaders'
                          ? LoaderWithUtils
                          : //
                          K extends keyof PropsFromConfig
                          ? PropsFromConfig[K]
                          : //
                            never;
                      } extends infer R
                      ? WithExtend<R>
                      : never
                    : never
                  : never
                : never
              : never
            : never
          : never
        : never
      : never
    : never;

export type _EntityMethods<Config extends EntityOptions> = WithExtend<
  PrimaryEntityMethods<Config>
>;

type WithExtend<T> = T extends {
  parse(...a: any[]): infer Doc;
}
  ? Doc extends DocumentBase
    ? {
        [K in
          | keyof T
          | 'extend'
          | 'addHooks'
          | 'addRelations']: K extends keyof T
          ? T[K]
          : //
          //
          // EXTEND
          K extends 'extend'
          ? <E>(
              transformer: (
                current: T,
                utils: { extend: <V>(value: V) => ExtendDefinitionResult<V, V> }
              ) => E
            ) => Omit<T, keyof E> & E extends infer R
              ? WithExtend<{ [K in keyof R]: R[K] } & {}>
              : never
          : //
          //
          //
          // ADD_HOOKS
          K extends 'addHooks'
          ? (
              options: MaybeArray<
                EntityHookOptions<Doc, EntityOptions['indexes']>
              >
            ) => WithExtend<T>
          : //
          //
          //
          //
          K extends 'addRelations'
          ? <
              Context extends LoaderContext,
              Definition extends ObjectFieldInput,
              ArgsDef extends ObjectDefinitionInput | undefined
            >(
              options: EntityFieldResolver<Context, Definition, ArgsDef, Doc>
            ) => WithExtend<T>
          : never;
      }
    : never
  : never;

export type EntityOperationInfoContext<
  Document extends DocumentBase,
  Indexes extends AnyCollectionIndexConfig['indexes'],
  Context extends LoaderContext
> = EntityOperationInfosRecord<
  Document,
  Indexes,
  Context
>[TransporterLoaderName];

type GetFieldsUsedInIndexes<IndexItem, Kind> = Kind extends keyof IndexItem
  ? IndexItem[Kind] extends Array<infer F> | ReadonlyArray<infer F>
    ? F extends `.${infer Field}`
      ? Field
      : never
    : never
  : never;

type EntityTransporterMethod<
  Method,
  Context extends LoaderContext = Record<string, any>
> = Method extends (config: infer Config) => infer Result
  ? ((
      config: Config & { context: Context } extends infer R
        ? {
            [K in keyof R as K extends 'context' ? never : K]: R[K];
          } & { context: Context }
        : never
    ) => Result) & {
      indexInfo: [ParsedIndexKey, ...ParsedIndexKey[]];
    }
  : never;

export type IndexMethods<
  Document extends DocumentBase,
  IndexItem,
  Context extends LoaderContext = Record<string, any>
> = {
  [M in keyof DocumentMethods<any, any, any>]: EntityTransporterMethod<
    DocumentMethods<
      Document,
      GetFieldsUsedInIndexes<IndexItem, 'PK'>,
      GetFieldsUsedInIndexes<IndexItem, 'SK'>
    >[M],
    Context
  >;
};

// type Index = {
//   name: 'a';
//   field: '_id';
//   PK: ['.name'];
//   SK: ['.userId'];
// };
// type Index2 = {
//   name: 'b';
//   field: '_id2';
//   PK: ['.userId'];
//   SK: ['.userId'];
// };
//
// type Doc = { name: string; userId: number; age?: number };
// type GT = {
//   definition: { def: { name: 'string'; userId: 'int'; age: 'int?' } };
// };
//
// type UPK = GetFieldsUsedInIndexes<Index, 'PK'>;
// type USK = GetFieldsUsedInIndexes<Index, 'SK'>;
// type M1 = DocumentMethods<Doc, UPK, USK>;
// type M2 = IndexMethods<Doc, Index, {}>;
// type OIM = OneIndexMethod<Doc, [Index, Index2]>;
// type LU = EntityLoadersWithUtils<GT, Doc, [Index, Index2]>;
//
// const x = {} as LU;
// type P = Parameters<typeof x.findOne>[0];
// type E = PromiseType<ReturnType<typeof x.findOne>>;
// const b = {} as P;
// b.filter.userId = { $eq: '' };
// const c = x.findOne.queryArgs.filter.def;

type GetIndexByName<
  Union extends DocumentIndexItem<any, any>,
  Name extends string
> = Union extends { [K: string]: unknown; name: Name } ? Union : never;

type OneIndexMethod<
  Doc extends DocumentBase,
  Indexes,
  Methods = {}
> = Indexes extends readonly [infer CurrentIndex, ...infer Rest]
  ? OneIndexMethod<
      Doc,
      Rest,
      {
        [K in keyof IndexMethods<Doc, CurrentIndex>]: K extends keyof Methods
          ? EntityTransporterMethod<
              IndexMethods<Doc, CurrentIndex>[K] extends (
                ...args: infer Args
              ) => infer Result
                ? (...args: Args) => Result
                : never
            > &
              EntityTransporterMethod<
                Methods[K] extends (...args: infer Args) => infer Result
                  ? (...args: Args) => Result
                  : never
              >
          : IndexMethods<Doc, CurrentIndex>[K];
      }
    >
  : [keyof Methods] extends [never]
  ? never
  : Methods;

type EntityLoaders<
  Doc extends DocumentBase,
  Indexes extends CollectionConfigIndexes<Doc>
> = UnionToIntersection<
  {
    [IndexName in Indexes[number]['name']]: {
      [M in keyof IndexMethods<any, any>]: {
        [L in `${M}${Capitalize<IndexName>}`]: IndexMethods<
          Doc,
          GetIndexByName<Indexes[number], IndexName>
        >[M];
      };
    };
  }[Indexes[number]['name']][Exclude<keyof IndexMethods<any, any>, 'createOne'>]
> &
  OneIndexMethod<Doc, Indexes> & {
    createOne: EntityTransporterMethod<
      DocumentMethods<
        //
        // Doc with DefaultEntityFields as optional
        {
          [K in keyof Doc as K extends keyof DefaultEntityFields
            ? never
            : K]: Doc[K];
        } & {
          [K in keyof Doc as K extends keyof DefaultEntityFields
            ? K
            : never]?: Doc[K];
        } extends infer R
          ? { [K in keyof R]: R[K] } & {}
          : never,
        // <--
        //
        GetFieldsUsedInIndexes<Indexes[number], 'PK'>,
        GetFieldsUsedInIndexes<Indexes[number], 'SK'>
      >['createOne'] extends infer CreateOne
        ? CreateOne extends (config: infer Config) => any
          ? (config: Config) => Promise<CreateOneResult<Doc>>
          : never
        : never
    >;
  };

type EntityLoadersWithUtils<
  Type,
  Doc extends DocumentBase,
  Indexes extends CollectionConfigIndexes<Doc>
> = EntityLoaders<Doc, Indexes> extends infer Loaders
  ? {
      [K in keyof Loaders]: _GetLoaderUtils<
        Loaders[K],
        Type
      > extends infer Utils
        ? [Utils] extends [never]
          ? Loaders[K]
          : Loaders[K] & Utils
        : never;
    }
  : never;

function _EntityGeneratedFields<
  T extends { [K in keyof DefaultEntityFields]: ObjectFieldInput }
>(
  input: T
): {
  [K in keyof T]: {
    [S in keyof ToFinalField<T[K]> as S extends '__infer'
      ? never
      : S]: ToFinalField<T[K]>[S];
  } & {};
} {
  return parseObjectDefinition(input).definition as any;
}

export type EntityDocument<Document> = {
  [K in keyof (DefaultEntityFields & Document)]: (DefaultEntityFields &
    Document)[K];
} & {};

export type AnyEntityDocument = EntityDocument<{ [K: string]: unknown }>;
