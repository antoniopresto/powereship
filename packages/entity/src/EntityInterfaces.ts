import {
  DescribeField,
  DescribeObjectDefinition,
  ExtendDefinitionResult,
  GraphType,
  Infer,
  MakeFieldOptional,
  ObjectDefinitionInput,
  ObjectFieldInput,
} from '@backland/schema';
import {
  CreateOne,
  DocumentBase,
  DocumentIndexesConfig,
  IndexMethods,
  LoaderContext,
  ParsedDocumentIndexes,
  ParsedIndexKey,
  Transporter,
  TransporterLoaderName,
  TransporterLoadersRecord,
} from '@backland/transporter';
import {
  DeleteMany,
  DeleteOne,
  FindById,
  FindMany,
  FindOne,
  Paginate,
  UpdateMany,
  UpdateOne,
} from '@backland/transporter';
import { Cast, Compute, GetFieldByDotNotation, Merge } from '@backland/utils';

import {
  EntityGraphQLConditionsType,
  EntityGraphQLFieldConditionsType,
} from './EntityFilterConditionType';
import {
  _EntityGraphType,
  EntityFieldResolver,
  EntityOptions,
} from './EntityOptions';
import { EntityHooks } from './EntityPlugin';
import {
  EntityDefaultFieldsDef,
  EntityOptionalDefaultFieldsDef,
} from './defaultFields';
import { EntityOperationInfosRecord } from './entityOperationContextTypes';
import { EntityIndexRelationConfig } from './indexRelations/addEntityIndexRelations';
import { EdgeType, PaginationType } from './paginationUtils';

export type GetLoaderFilterDef<LoaderConfig, DocDef> =
  //
  LoaderConfig extends { filter: infer Filter }
    ? {
        [K in keyof Filter as K extends keyof DocDef
          ? K
          : never]: K extends keyof DocDef
          ? Omit<DescribeField<DocDef[K]>, 'optional'> & {
              optional: true;
            }
          : never;
      }
    : never;

export type EntityDefaultFields = {
  _v: string;
  createdAt: Date;
  createdBy: string | undefined;
  id: string;
  ulid: string;
  updatedAt: Date;
  updatedBy: string | undefined;
};

export type EntityUtils<
  LoaderConfig,
  OutputDefinition extends ObjectDefinitionInput,
  FilterDef extends GetLoaderFilterDef<
    LoaderConfig,
    OutputDefinition
  > = GetLoaderFilterDef<LoaderConfig, OutputDefinition>
> = {
  indexInfo: [ParsedIndexKey, ...ParsedIndexKey[]];
  filterDef: Compute<FilterDef>;
  queryArgs: {
    after: 'ID?';
    condition: {
      object: EntityGraphQLFieldConditionsType<OutputDefinition>;
      optional: true;
    };
    filter: { type: 'object'; def: Compute<FilterDef> };
    first: {
      optional: true;
      type: 'int';
    };
  };
};

export type _EntityWithUtils<
  Loader,
  OutputDefinition extends ObjectDefinitionInput
> = Loader extends (config: infer Config) => infer Res
  ? ((config: Config) => Res) & EntityUtils<Config, OutputDefinition>
  : Loader;

export type EntityOutputDoc<Input extends ObjectDefinitionInput> = Cast<
  Compute<Merge<EntityDefaultFields, Infer<{ object: Input }>>>,
  Record<string, any>
>;

export type EntityInputDoc<Input extends ObjectDefinitionInput> = Cast<
  Compute<Merge<Partial<EntityDefaultFields>, Infer<{ object: Input }>>>,
  Record<string, any>
>;

export type EntityInputDef<Input extends ObjectDefinitionInput> = Cast<
  Merge<EntityOptionalDefaultFieldsDef, Input>,
  ObjectDefinitionInput
>;

export interface Entity<
  Input extends ObjectDefinitionInput,
  Indexes extends DocumentIndexesConfig
> {
  name: string;
  usedOptions: EntityOptions<Input, Indexes>;
  inputConfigTypeDefinition: Input;
  indexes: Indexes;
  outputDefinition: Cast<
    Merge<EntityDefaultFieldsDef, Input>,
    ObjectDefinitionInput
  >;
  type: GraphType<{
    object: Merge<EntityDefaultFieldsDef, Input>;
  }>;
  inputType: GraphType<{
    object: EntityInputDef<Input>;
  }>;
  createOne: CreateOne<EntityInputDoc<Input>, EntityOutputDoc<Input>, Indexes>;
  findOne: _EntityWithUtils<
    FindOne<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  findMany: _EntityWithUtils<
    FindMany<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  paginate: _EntityWithUtils<
    Paginate<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  deleteMany: _EntityWithUtils<
    DeleteMany<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  deleteOne: _EntityWithUtils<
    DeleteOne<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  findById: _EntityWithUtils<
    FindById<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  updateMany: _EntityWithUtils<
    UpdateMany<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;
  updateOne: _EntityWithUtils<
    UpdateOne<EntityOutputDoc<Input>, Indexes>,
    this['outputDefinition']
  >;

  addIndexRelation: <E extends unknown, Name extends string>(
    name: Name,
    entity: E
  ) => Entity<
    Omit<Input, Name> & {
      [L in Name]: { array: { of: GetFieldByDotNotation<E, 'inputType'> } };
    },
    Indexes
  >;

  aliasPaths: string[];

  conditionsDefinition: {
    def: EntityGraphQLConditionsType<Input>;
    type: 'object';
  };

  databaseType: this['type'];

  edgeType: EdgeType<this['type']>;

  extendType: <T extends _EntityGraphType>(
    handler: (
      helper: ExtendDefinitionResult<this['inputType'], this['inputType']>,
      originalOptions: this['usedOptions']
    ) => T
  ) => Entity<
    T['definition']['def'] extends ObjectDefinitionInput
      ? T['definition']['def']
      : {},
    Indexes
  >;

  getDocumentId(doc: Record<string, any>): string;

  readonly hasAliases: boolean;

  indexGraphTypes: {
    [K: string]: GraphType<{
      object: ObjectDefinitionInput;
    }>;
  };

  // paths of found aliases in entity schemas or sub schemas
  indexRelations: { [K: string]: EntityIndexRelationConfig };

  originType: GraphType<{ object: Input }>;

  paginationType: PaginationType<this['type']>;

  parse: (...args: Parameters<this['type']['parse']>) => EntityOutputDoc<Input>;

  parseDocumentIndexes(doc: Record<string, any>): ParsedDocumentIndexes;

  setOption: <Key extends keyof this['usedOptions'], V>(
    optionName: Key,
    value: V
  ) => this;

  transporter: Transporter | undefined;

  updateDefinition: Compute<
    MakeFieldOptional<DescribeObjectDefinition<Input>, keyof Input>
  >;

  addHooks: (options: (hooks: EntityHooks) => any) => this;

  addRelation: <
    Context extends LoaderContext,
    Definition extends ObjectFieldInput,
    ArgsDef extends ObjectDefinitionInput
  >(
    options: EntityFieldResolver<
      Context,
      Definition,
      ArgsDef,
      EntityOutputDoc<Input>
    >
  ) => this;

  extend: <TransformerReturn>(
    transformer: (
      current: ExcludeExtend<this>,
      utils: {
        extend: <V>(value: V) => ExtendDefinitionResult<V, V>;
      }
    ) => TransformerReturn
  ) => this extends infer Origin
    ? {
        [K in
          | keyof TransformerReturn
          | keyof Origin]: K extends keyof TransformerReturn
          ? TransformerReturn[K]
          : K extends keyof Origin
          ? Origin[K]
          : never;
      }
    : never;

  hooks: EntityHooks;
  __$is_entity__: true;
}

type ExtendMethodKeys = 'addHooks' | 'addRelation' | 'extend';

type ExcludeExtend<E> = {
  [K in keyof E as K extends ExtendMethodKeys ? never : K]: E[K];
} & {};

export type EntityOperationInfoContext<
  LoaderName extends TransporterLoaderName = TransporterLoaderName
> = EntityOperationInfosRecord[LoaderName];

export function isEntityContextOfLoader<
  LoaderName extends TransporterLoaderName
>(
  t: EntityOperationInfoContext,
  name: LoaderName
): t is EntityOperationInfoContext<LoaderName> {
  return t.loaderName === name;
}

export type EntityLoaderConfig<
  Method extends TransporterLoaderName,
  Context extends LoaderContext = Record<string, any>
> = TransporterLoadersRecord[Method] extends (config: infer Config) => any
  ? Config & { context: Context } extends infer R
    ? {
        [K in keyof R as K extends 'context' ? never : K]: R[K];
      } & { context: Context }
    : never
  : any;

export type EntityDocument<Document extends DocumentBase = DocumentBase> =
  Compute<Merge<EntityDefaultFields, Document>>;

export type AnyEntityDocument<Doc extends DocumentBase = DocumentBase> =
  EntityDocument<Doc>;

export type DocumentDefinitionAsLiterals<Doc> = {
  [K in keyof Doc]: {
    literal: Doc[K];
  };
} & {};

export interface _AnyEntity<Doc extends DocumentBase = DocumentBase>
  extends Entity<
    DocumentDefinitionAsLiterals<Omit<Doc, keyof EntityDefaultFields>>,
    DocumentIndexesConfig<keyof Omit<Doc, keyof EntityDefaultFields>>
  > {
  //
}
export interface AnyEntity<Doc extends DocumentBase = DocumentBase>
  extends Omit<_AnyEntity<Doc>, TransporterLoaderName>,
    IndexMethods<Doc, DocumentIndexesConfig> {
  //
}
