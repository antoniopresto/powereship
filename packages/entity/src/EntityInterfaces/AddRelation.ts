import { ObjectDefinitionInput, ObjectFieldInput } from '@backland/schema';
import { LoaderContext } from '@backland/transporter';

import { EntityFieldResolver } from '../EntityOptions';

import { AnyEntityTypesContext } from './Context';

export interface EntityAddRelation<
  Parent,
  EContext extends AnyEntityTypesContext
> {
  <
    Context extends LoaderContext,
    Definition extends ObjectFieldInput,
    ArgsDef extends ObjectDefinitionInput
  >(
    options: EntityFieldResolver<
      Context,
      Definition,
      ArgsDef,
      EContext['document']
    >
  ): Parent;
}
