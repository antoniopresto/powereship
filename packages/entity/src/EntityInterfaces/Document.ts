import { IsKnown, Merge } from '@powership/utils';

export type EntityDocumentBase = {
  _v: string;
  _c: string;
  createdAt: Date;
  createdBy: string | undefined;
  id: string;
  ulid: string;
  updatedAt: Date;
  updatedBy: string | undefined;
};

export type EntityDocument<Shape> = IsKnown<Shape> extends 1
  ? Merge<EntityDocumentBase, Shape>
  : EntityDocumentBase & { [K: string]: unknown };

export type EntityDocumentInput<Shape = {}> = IsKnown<Shape> extends 1
  ? Merge<Partial<EntityDocumentBase>, Shape>
  : Partial<EntityDocumentBase> & { [K: string]: unknown };
