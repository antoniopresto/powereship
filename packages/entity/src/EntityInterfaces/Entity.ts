import { DocumentIndexItem } from '@powership/transporter';

import { EntityTypesContext } from './Context';
import { EntityFromContext } from './EntityFromContext';

export interface Entity<Input, Indexes>
  extends EntityFromContext<EntityTypesContext<Input, Indexes>> {
  name: string;
}

export type _AnyEntity = EntityFromContext<
  EntityTypesContext<{}, DocumentIndexItem[]>
>;

export type AnyEntity = {
  [K in keyof _AnyEntity]: any;
};
