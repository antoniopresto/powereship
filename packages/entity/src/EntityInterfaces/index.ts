export * from './AddRelation';
export * from './Context';
export * from './Document';
export * from './Entity';
export * from './EntityFromContext';
export * from './EntityLoaderConfig';
export * from './EntityLoaderMethods';
export * from './ExtendEntity';
export * from './ExtractUtils';
