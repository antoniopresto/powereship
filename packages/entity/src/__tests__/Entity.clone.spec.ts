import { setupProductTest } from './setupProductTest';
import { assert, IsExact } from 'conditional-type-checks';
import { EntityDefaultFields } from '../EntityInterfaces';

describe('Entity clone', () => {
  const { getMocks } = setupProductTest();

  test('cloning', async function () {
    const { ProductEntity } = await getMocks();

    const clonedEntity = ProductEntity.clone((config) => {
      return {
        ...config,
        name: 'Cloned',
        indexes: [
          {
            name: 'sku',
            PK: ['.sku'],
            field: '_id',
          },
        ],
        type: config.type.clone().only(['sku']).graphType('ClonedType'),
      };
    });

    type Cloned = ReturnType<typeof clonedEntity.parse>;
    assert<IsExact<Cloned, EntityDefaultFields & { sku: string }>>(true);

    let productGQL = ProductEntity.type.print();

    expect(clonedEntity.type.print()).toEqual([
      'type ClonedEntity {',
      '  createdAt: Date!',
      '  createdBy: String',
      '  id: String!',
      '  ulid: Ulid!',
      '  updatedAt: Date!',
      '  updatedBy: String',
      '  sku: String!',
      '}',
      '',
      'scalar Date',
      '',
      'scalar Ulid',
      '',
      'input ClonedEntityInput {',
      '  createdAt: Date!',
      '  createdBy: String',
      '  id: String!',
      '  ulid: Ulid!',
      '  updatedAt: Date!',
      '  updatedBy: String',
      '  sku: String!',
      '}',
    ]);

    expect(productGQL).toMatchSnapshot();
  });
});
