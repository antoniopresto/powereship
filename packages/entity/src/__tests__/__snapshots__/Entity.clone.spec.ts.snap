// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity clone cloning 1`] = `
[
  "type ProductEntity {",
  "  createdAt: Date!",
  "  createdBy: String",
  "  id: String!",
  "  ulid: Ulid!",
  "  updatedAt: Date!",
  "  updatedBy: String",
  "  sku: String!",
  "  storeId: ID!",
  "  title: String!",
  "  stock: Stock!",
  "  shortDescription: String",
  "  brand: String!",
  "  detailsUrl: String",
  "  alcoholic: Boolean!",
  "  thumbUrl: String",
  "  breadcrumb: [BreadCrumb]",
  "  mapOfImages: [ProductImageMap]",
  "  attributes: ProductEntity_attributes",
  "  currentPrice: Float!",
  "  priceFrom: Float",
  "  sellPrice: Float!",
  "  dimensions: Dimensions",
  "  tags: [String]",
  "  slug: String",
  "  mainCategoryId: ID",
  "  categories: [String]!",
  "  spotlight: Boolean",
  "  html: String",
  "}",
  "",
  "scalar Date",
  "",
  "scalar Ulid",
  "",
  "type Stock {",
  "  available: Boolean!",
  "  count: Float",
  "  maxCartQty: Float",
  "",
  "  """Should track count"""",
  "  track: Boolean",
  "}",
  "",
  "type BreadCrumb {",
  "  id: ID!",
  "  active: Boolean",
  "  name: String!",
  "  parentId: ID",
  "}",
  "",
  "type ProductImageMap {",
  "  key: String",
  "  kind: String",
  "  allowZoom: Boolean",
  "}",
  "",
  "scalar ProductEntity_attributes",
  "",
  "type Dimensions {",
  "  weight: String",
  "  length: String",
  "  height: String",
  "  width: String",
  "}",
  "",
  "input ProductEntityInput {",
  "  createdAt: Date!",
  "  createdBy: String",
  "  id: String!",
  "  ulid: Ulid!",
  "  updatedAt: Date!",
  "  updatedBy: String",
  "  sku: String!",
  "  storeId: ID!",
  "  title: String!",
  "  stock: StockInput!",
  "  shortDescription: String",
  "  brand: String!",
  "  detailsUrl: String",
  "  alcoholic: Boolean!",
  "  thumbUrl: String",
  "  breadcrumb: [BreadCrumbInput]",
  "  mapOfImages: [ProductImageMapInput]",
  "  attributes: ProductEntity_attributes",
  "  currentPrice: Float!",
  "  priceFrom: Float",
  "  sellPrice: Float!",
  "  dimensions: DimensionsInput",
  "  tags: [String]",
  "  slug: String",
  "  mainCategoryId: ID",
  "  categories: [String]!",
  "  spotlight: Boolean",
  "  html: String",
  "}",
  "",
  "input StockInput {",
  "  available: Boolean!",
  "  count: Float",
  "  maxCartQty: Float",
  "",
  "  """Should track count"""",
  "  track: Boolean",
  "}",
  "",
  "input BreadCrumbInput {",
  "  id: ID!",
  "  active: Boolean",
  "  name: String!",
  "  parentId: ID",
  "}",
  "",
  "input ProductImageMapInput {",
  "  key: String",
  "  kind: String",
  "  allowZoom: Boolean",
  "}",
  "",
  "input DimensionsInput {",
  "  weight: String",
  "  length: String",
  "  height: String",
  "  width: String",
  "}",
]
`;

exports[`Entity clone setOption 1`] = `
[
  "type ProductEntity {",
  "  createdAt: Date!",
  "  createdBy: String",
  "  id: String!",
  "  ulid: Ulid!",
  "  updatedAt: Date!",
  "  updatedBy: String",
  "  sku: String!",
  "  storeId: ID!",
  "  title: String!",
  "  stock: Stock!",
  "  shortDescription: String",
  "  brand: String!",
  "  detailsUrl: String",
  "  alcoholic: Boolean!",
  "  thumbUrl: String",
  "  breadcrumb: [BreadCrumb]",
  "  mapOfImages: [ProductImageMap]",
  "  attributes: ProductEntity_attributes",
  "  currentPrice: Float!",
  "  priceFrom: Float",
  "  sellPrice: Float!",
  "  dimensions: Dimensions",
  "  tags: [String]",
  "  slug: String",
  "  mainCategoryId: ID",
  "  categories: [String]!",
  "  spotlight: Boolean",
  "  html: String",
  "}",
  "",
  "scalar Date",
  "",
  "scalar Ulid",
  "",
  "type Stock {",
  "  available: Boolean!",
  "  count: Float",
  "  maxCartQty: Float",
  "",
  "  """Should track count"""",
  "  track: Boolean",
  "}",
  "",
  "type BreadCrumb {",
  "  id: ID!",
  "  active: Boolean",
  "  name: String!",
  "  parentId: ID",
  "}",
  "",
  "type ProductImageMap {",
  "  key: String",
  "  kind: String",
  "  allowZoom: Boolean",
  "}",
  "",
  "scalar ProductEntity_attributes",
  "",
  "type Dimensions {",
  "  weight: String",
  "  length: String",
  "  height: String",
  "  width: String",
  "}",
  "",
  "input ProductEntityInput {",
  "  createdAt: Date!",
  "  createdBy: String",
  "  id: String!",
  "  ulid: Ulid!",
  "  updatedAt: Date!",
  "  updatedBy: String",
  "  sku: String!",
  "  storeId: ID!",
  "  title: String!",
  "  stock: StockInput!",
  "  shortDescription: String",
  "  brand: String!",
  "  detailsUrl: String",
  "  alcoholic: Boolean!",
  "  thumbUrl: String",
  "  breadcrumb: [BreadCrumbInput]",
  "  mapOfImages: [ProductImageMapInput]",
  "  attributes: ProductEntity_attributes",
  "  currentPrice: Float!",
  "  priceFrom: Float",
  "  sellPrice: Float!",
  "  dimensions: DimensionsInput",
  "  tags: [String]",
  "  slug: String",
  "  mainCategoryId: ID",
  "  categories: [String]!",
  "  spotlight: Boolean",
  "  html: String",
  "}",
  "",
  "input StockInput {",
  "  available: Boolean!",
  "  count: Float",
  "  maxCartQty: Float",
  "",
  "  """Should track count"""",
  "  track: Boolean",
  "}",
  "",
  "input BreadCrumbInput {",
  "  id: ID!",
  "  active: Boolean",
  "  name: String!",
  "  parentId: ID",
  "}",
  "",
  "input ProductImageMapInput {",
  "  key: String",
  "  kind: String",
  "  allowZoom: Boolean",
  "}",
  "",
  "input DimensionsInput {",
  "  weight: String",
  "  length: String",
  "  height: String",
  "  width: String",
  "}",
]
`;
